internal/cli/internal/command/ssh/ssh_terminal.go

```diff
@@ -0,0 +1,204 @@
+package ssh
+
+import (
+	"context"
+	"crypto/ed25519"
+	"encoding/base64"
+	"encoding/pem"
+	"fmt"
+	"io"
+	"math/rand"
+	"os"
+	"time"
+
+	sshCrypt "golang.org/x/crypto/ssh"
+
+	"github.com/briandowns/spinner"
+	"github.com/pkg/errors"
+	"github.com/superfly/flyctl/api"
+	"github.com/superfly/flyctl/helpers"
+	"github.com/superfly/flyctl/internal/client"
+	"github.com/superfly/flyctl/pkg/agent"
+	"github.com/superfly/flyctl/pkg/ssh"
+	"github.com/superfly/flyctl/terminal"
+)
+
+func spin(in, out string) context.CancelFunc {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	if !helpers.IsTerminal() {
+		fmt.Fprintln(os.Stderr, in)
+		return cancel
+	}
+
+	go func() {
+		s := spinner.New(spinner.CharSets[11], 100*time.Millisecond)
+		s.Writer = os.Stderr
+		s.Prefix = in
+		s.FinalMSG = out
+		s.Start()
+		defer s.Stop()
+
+		<-ctx.Done()
+	}()
+
+	return cancel
+}
+
+type SSHParams struct {
+	Ctx            context.Context
+	Org            *api.Organization
+	App            string
+	Dialer         agent.Dialer
+	Cmd            string
+	Stdin          io.Reader
+	Stdout         io.WriteCloser
+	Stderr         io.WriteCloser
+	DisableSpinner bool
+}
+
+func SSHConnect(p *SSHParams, addr string) error {
+	terminal.Debugf("Fetching certificate for %s\n", addr)
+
+	cert, err := singleUseSSHCertificate(p.Ctx, p.Org)
+	if err != nil {
+		return fmt.Errorf("create ssh certificate: %w (if you haven't created a key for your org yet, try `flyctl ssh establish`)", err)
+	}
+
+	pk, err := parsePrivateKey(cert.Key)
+	if err != nil {
+		return errors.Wrap(err, "parse ssh certificate")
+	}
+
+	pemkey := MarshalED25519PrivateKey(pk, "single-use certificate")
+
+	terminal.Debugf("Keys for %s configured; connecting...\n", addr)
+
+	sshClient := &ssh.Client{
+		Addr: addr + ":22",
+		User: "root",
+
+		Dial: p.Dialer.DialContext,
+
+		Certificate: cert.Certificate,
+		PrivateKey:  string(pemkey),
+	}
+
+	var endSpin context.CancelFunc
+	if !p.DisableSpinner {
+		endSpin = spin(fmt.Sprintf("Connecting to %s...", addr),
+			fmt.Sprintf("Connecting to %s... complete\n", addr))
+		defer endSpin()
+	}
+
+	if err := sshClient.Connect(context.Background()); err != nil {
+		return errors.Wrap(err, "error connecting to SSH server")
+	}
+	defer sshClient.Close()
+
+	terminal.Debugf("Connection completed.\n", addr)
+
+	if !p.DisableSpinner {
+		endSpin()
+	}
+
+	term := &ssh.Terminal{
+		Stdin:  p.Stdin,
+		Stdout: p.Stdout,
+		Stderr: p.Stderr,
+		Mode:   "xterm",
+	}
+
+	if err := sshClient.Shell(context.Background(), term, p.Cmd); err != nil {
+		return errors.Wrap(err, "ssh shell")
+	}
+
+	return nil
+}
+
+// // stolen from `mikesmitty`, thanks, you are a mikesmitty and a scholar
+func MarshalED25519PrivateKey(key ed25519.PrivateKey, comment string) []byte {
+	magic := append([]byte("openssh-key-v1"), 0)
```

Comment:
this can easily be `magic := []byte("openssh-key-v1\x00")`

Reply:
Got it, this entire file was copy and pasted over from the `cmd` package.